File: ruby/linked-list/linked_list.rb
class Deque
  private

  attr_reader :head

  class Node
    attr_accessor :pred, :succ
    attr_reader :datum

    def initialize(datum, pred, succ)
      @datum = datum
      @pred = pred || self
      @succ = succ || self
    end

    def unlink
      pred.succ = succ
      succ.pred = pred
      self
    end

    def link
      pred.succ = self
      succ.pred = self
      self
    end
  end

  public

  def initialize
    @head = Node.new(:sentinel, nil, nil)
  end

  def unshift(datum)
    Node.new(datum, head, head.succ).link
    self
  end

  def shift
    head.succ.unlink.datum
  end

  def push(datum)
    Node.new(datum, head.pred, head).link
    self
  end

  def pop
    head.pred.unlink.datum
  end

end

File: python/linked-list/linked_list.py
class Node(object):
    def __init__(self, value, next=None, prev=None):
        self.value = value
        self.next = next
        self.prev = prev


class LinkedList:
    def __init__(self):
        self.head = Node(None)
        self.head.next = self.head.prev = self.head

    def _unlink(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        return node.value

    def _link(self, value, prev):
        node = Node(value, prev.next, prev)
        node.prev.next = node
        node.next.prev = node

    def shift(self):
        return self._unlink(self.head.next)

    def unshift(self, value):
        self._link(value, self.head)

    def pop(self):
        return self._unlink(self.head.prev)

    def push(self, value):
        self._link(value, self.head.prev)
